bundles/org.openhab.binding.nibeheatpump/src/test/java/org/openhab/binding/nibeheatpump/internal/handler/NibeHeatPumpHandlerNibe2StateTest.java

```diff
@@ -0,0 +1,110 @@
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.nibeheatpump.internal.handler;
+
+import org.eclipse.smarthome.core.types.State;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.openhab.binding.nibeheatpump.internal.models.PumpModel;
+import org.openhab.binding.nibeheatpump.internal.models.VariableInformation;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.Arrays;
+import java.util.Collection;
+
+import static org.junit.Assert.assertEquals;
+
+/**
+ * Tests cases for {@link NibeHeatPumpHandler}.
+ *
+ * @author Jevgeni Kiski - Initial contribution
+ */
+@RunWith(Parameterized.class)
+public class NibeHeatPumpHandlerNibe2StateTest {
+    private NibeHeatPumpHandler product; // the class under test
+    private Method m;
+    private static String METHOD_NAME = "convertNibeValueToState";
+    private Class[] parameterTypes;
+    private Object[] parameters;
+
+    private int fCoilAddress;
+
+    private int fValue;
+
+    private String fFormat;
+
+    private String fType;
+
+    private String fExpected;
+
+    @Parameterized.Parameters(name = "{index}: f({0}, {1}, {3})={4}")
+    public static Collection<Object[]> data() {
+        return Arrays.asList(new Object[][] {
+                { 47028, 0xFF, "%d", "Number", "-1" },
+                { 48132, 1966080, "%s", "String", "0" },
+                { 47028, (byte)0xFF, "%d", "Number", "-1" },
+                { 43009, 0x011F, "%.1f", "Number", "28.7" },
+                { 40004, 0xFFFF, "%.1f", "Number", "-0.1" },
+                { 40004, (short)0xFFFF, "%.1f", "Number", "-0.1" },
+                { 43416, 0xFFFFFFFF, "%d", "Number", "4294967295" },
+                { 47418, 0x004B, "%d", "Number", "75" },
+                { 43514, 0x0007, "%d", "Number", "7" },
+                { 47291, 0xFFFF, "%d", "Number", "65535" },
+                { 42437, 0xFFFFFFFF, "%.1f", "Number", "429496729.5" },
+                { 42504, 0xFFFFFFFF, "%d", "Number", "4294967295" },
+                { 43081, 196, "%.1f", "Number", "19.6"},
+                { 43424, 1685, "%d", "Number", "1685"},
+                { 43416, 4857, "%d", "Number", "4857"},
+                { 43420, 9487, "%d", "Number", "9487"},
+                { 40940, (byte)0xFF, "%.1f", "Number", "-0.1"}, // these tests pass, but is it good?
+                { 40940, (int)((short)0xFFFF), "%.1f", "Number", "-0.1"}, // these tests pass, but is it good?
+                { 40940, (short)0xFFFF, "%.1f", "Number", "-0.1"}, // these tests pass, but is it good?
+                { 40940, (byte)0xFFFF, "%.1f", "Number", "-0.1"}, // these tests pass, but is it good?
+                { 40940, (int)(0xFFFF), "%.1f", "Number", "6553.5"}, // these tests pass, but is it good?
```

Comment:
With simple value convert approach I do not like this: Why produced result depends on what was initial cast?

Reply:
For S32:
This might be a solution:
`x = (int)((long)value << 1) >>> 1;`

Any ideas?